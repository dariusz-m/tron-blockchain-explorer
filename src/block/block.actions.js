export const BlockActionTypes = {
    LOAD_BLOCK: "LOAD_BLOCK",
    BLOCK_HAS_BEEN_LOADED: "BLOCK_HAS_BEEN_LOADED",
    BLOCK_HAS_NOT_BEEN_LOADED: "BLOCK_HAS_NOT_BEEN_LOADED",
};

export const loadBlock = () => {
    return {
        type: BlockActionTypes.LOAD_BLOCK
    };
};

export const blockHasBeenLoaded = (block) => {
    return {
        type: BlockActionTypes.BLOCK_HAS_BEEN_LOADED,
        block
    };
};

export const blockHasNotBeenLoaded = () => {
    return {
        type: BlockActionTypes.BLOCK_HAS_NOT_BEEN_LOADED,
    };
};
