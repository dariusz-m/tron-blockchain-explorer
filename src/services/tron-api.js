import HttpClient from "@tronprotocol/wallet-api/src/client/http";

export const TronBlockchainClient = new HttpClient();
