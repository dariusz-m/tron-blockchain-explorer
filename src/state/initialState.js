const initialState = {
    router: {"location": null},
    tokens: [],
    selectedBlock: {number: -1, witnessId: -1, hash: "", parentHash: ""}
};

export default initialState;
