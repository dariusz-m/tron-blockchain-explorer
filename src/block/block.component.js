import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import {bindActionCreators} from "redux";

import {loadBlock} from "./block.actions";

export class Block extends React.Component {
    componentDidMount() {
        this.props.actions.loadBlock();
    }

    render() {
        return (
            <div>
                <h2>Block</h2>
                <div><pre>{JSON.stringify(this.props.selectedBlock, null, 2) }</pre></div>
            </div>
        );
    }
}

Block.propTypes = {
    selectedBlock: PropTypes.shape({
        number: PropTypes.number.isRequired,
        witnessId: PropTypes.number.isRequired,
        hash: PropTypes.string.isRequired,
        parentHash: PropTypes.string.isRequired,
    }).isRequired,
    actions: PropTypes.shape({
        loadBlock: PropTypes.func.isRequired,
    }),
};

function mapStateToProps(state) {
    return {
        selectedBlock: state.selectedBlock
    };
}

function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators({
            loadBlock: loadBlock,
        }, dispatch)
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(Block);
