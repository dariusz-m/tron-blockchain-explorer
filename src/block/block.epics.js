import { from, of, merge } from 'rxjs';
import { mergeMap, map, catchError, pluck} from 'rxjs/operators';
import { ajax } from 'rxjs/ajax';

import {
    BlockActionTypes,
    blockHasBeenLoaded,
    blockHasNotBeenLoaded,
    transactionsByBlockHaveBeenLoaded, transactionsByBlockHaveNotBeenLoaded
} from './block.actions';
import {makeRequest, TYPE_OF_REQUESTS} from "../search/search.epics";
import {filter} from "rxjs/operators/index";
import {TRON_ORGIN} from "../config/tron-api.config";

export const loadBlock = action$ => {
    return action$.ofType(BlockActionTypes.LOAD_BLOCK)
        .pipe(
            mergeMap((action) => merge(
                makeRequest(
                    `${TRON_ORGIN}/api/block?sort=-number&limit=1&number=${action.payload}`,
                    TYPE_OF_REQUESTS.BLOCK_BY_NUMBER
                ),
                makeRequest(
                    `${TRON_ORGIN}/api/block?sort=-number&limit=1&hash=${action.payload}`,
                    TYPE_OF_REQUESTS.BLOCK_BY_HASH
                )
            ).pipe(
                filter((response) => response.data.length > 0),
                map((response) => blockHasBeenLoaded(response.data[0])),
                catchError((error) => {
                    console.error(error);
                    return of(blockHasNotBeenLoaded());
                }),
            )),
        );
};

export const loadTransactionsByBlock = action$ =>
    action$.ofType(BlockActionTypes.LOAD_TRANSACTIONS_BY_BLOCK)
        .pipe(
            mergeMap((action) =>
                from(
                    ajax(`${TRON_ORGIN}/api/transaction?sort=-timestamp&limit=100&start=0&block=${action.payload}`)
                ).pipe(
                    pluck('response','data'),
                    map((transactions) => transactionsByBlockHaveBeenLoaded(transactions)),
                    catchError((error) => {
                        console.error(error);
                        return of(transactionsByBlockHaveNotBeenLoaded());
                    })
                )
            ),
        );
