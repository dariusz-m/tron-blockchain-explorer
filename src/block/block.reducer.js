import initialState from '../state/initialState';
import {BlockActionTypes} from './block.actions';

export default (state = initialState.selectedBlock, action) => {
    switch (action.type) {
        case BlockActionTypes.BLOCK_HAS_BEEN_LOADED:
            return {...action.block};
        default:
            return state;
    }
};
