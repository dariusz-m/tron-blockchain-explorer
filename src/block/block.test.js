import React from 'react';
import renderer from 'react-test-renderer';

import {Block} from "./block.component";
import configureStore from "../state";
import {loadBlock} from "./block.actions";
import initialState from "../state/initialState";
import {waitForAsynchronousActions} from '../tests.support';

jest.mock('../services/tron-api');
import {TronBlockchainClient} from '../services/tron-api';

describe('Block tests', () => {
    let block = {};

    beforeEach(() => {
        block = {
            "number": 145518,
            "witnessId": 0,
            "hash": "831906F5AED9F8C27A02AFC283BE3F5D3977C7F3D3253EB6BE448373C0708382",
            "parentHash": "000000000002386D3AADF9A5E5C34D0DFE007E4D4184788CECDF05DBDBCF1CF2"
        };
    });

    it('Renders correctly', () => {
        const actions = {
            loadBlock: () => {},
        };
        const tree = renderer.create(<Block actions={actions} selectedBlock={block}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Block has been loaded', async () => {
        TronBlockchainClient.getLatestBlock.mockImplementationOnce(async () => await {...block});
        const store = configureStore();
        const expectedState = {...initialState, selectedBlock: {...block}};

        store.dispatch(loadBlock());

        await waitForAsynchronousActions().then(() => expect(store.getState()).toEqual(expectedState));
    });
});

