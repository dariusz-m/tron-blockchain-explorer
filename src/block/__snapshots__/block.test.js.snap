// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block tests Renders correctly 1`] = `
<div>
  <h2>
    Block
  </h2>
  <div>
    <pre>
      {
  "number": 145518,
  "witnessId": 0,
  "hash": "831906F5AED9F8C27A02AFC283BE3F5D3977C7F3D3253EB6BE448373C0708382",
  "parentHash": "000000000002386D3AADF9A5E5C34D0DFE007E4D4184788CECDF05DBDBCF1CF2"
}
    </pre>
  </div>
</div>
`;
